sycl/doc/extensions/GroupMask/GroupMask.asciidoc

```diff
@@ -0,0 +1,341 @@
+
+This document describes an extension which adds a `group_mask` type.  Such a mask can be used to efficiently represent subsets of work-items in a group for which a given Boolean condition holds.  Group mask functionality is currently limited to groups that are instances of the `sub_group` class.
+
+== Notice
+
+Copyright (c) 2021 Intel Corporation.  All rights reserved.
+
+== Status
+
+Working Draft
+
+This is a preview extension specification, intended to provide early access to a feature for review and community feedback. When the feature matures, this specification may be released as a formal extension.
+
+Because the interfaces defined by this specification are not final and are subject to change they are not intended to be used by shipping software products.
+
+== Version
+
+Revision: 1
+
+== Contact
+John Pennycook, Intel (john 'dot' pennycook 'at' intel 'dot' com)
+
+== Dependencies
+
+This extension is written against the SYCL 2020 specification, Revision 3.
+
+== Feature Test Macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification section 6.3.3 "Feature test macros".  Therefore, an
+implementation supporting this extension must predefine the macro
+`SYCL_EXT_ONEAPI_GROUP_MASK` to one of the values defined in the table below.
+Applications can test for the existence of this macro to determine if the
+implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's APIs the implementation supports.
+
+[%header,cols="1,5"]
+|===
+|Value |Description
+|1     |Initial extension version.  Base features are supported.
+|===
+
+== Overview
+
+A group mask is an integral type sized such that each work-item in the group is
+represented by a single bit. Such a mask can be used to efficiently represent
+subsets of work-items in a group for which a given Boolean condition holds.
+
+Group mask functionality is currently limited to groups that are instances of
+the `sub_group` class, but this limitation may be lifted in a future version of
+the specification.
+
+=== Ballot
+
+The `group_ballot` algorithm converts a Boolean condition from each work-item
+in the group into a group mask.  Like other group algorithms, `group_ballot`
+must be encountered by all work-items in the group in converged control flow.
+
+|===
+|Function|Description
+
+|`template <typename Group> Group::mask_type group_ballot(Group g, bool predicate = true) const`
+|Return a `group_mask` representing the set of work-items in group _g_ for which _predicate_ is `true`.
+|===
+
+=== Group Masks
+
+The group mask type is an opaque type, permitting implementations to use any
+mask representation subject to the following restrictions:
+
+- The size and alignment of the mask type must be the same on the host and
+  device
+- A SYCL implementation supporting OpenCL interoperability must use a 128-bit
+  mask convertible to a `vec<uint,4>`
```

Comment:
It's not clear to me what this means exactly.  Must there be an implicit conversion from `group_mask` to `vec<uint,4>`?

It might make more sense to provide an OpenCL backend specific function like:

```
namespace sycl::opencl {

vec<uint,4> group_mask_to_vec(const group_mask &gm);

}
```

Reply:
I agree that this restriction is a little weird; I think if we were designing this extension today we might have done it differently.  Thanks for pointing it out.

This limitation actually comes from SPIR-V (see the documentation for [OpGroupNonUniformBallot](https://www.khronos.org/registry/SPIR-V/specs/unified1/SPIRV.html#OpGroupNonUniformBallot) and similar functions).  I suppose what this restriction is really trying to say is that if an implementation relies on these SPIR-V instructions to implement the underlying functionality, an implementation is effectively limited to 128-bit masks -- and that such an implementation would rely on being able to represent that mask as a vector of 32-bit integers.  Part of the worry here, I think, was that somebody might try to get a mask and somehow call an OpenCL or SPIR-V intrinsic with it.

I think what you're proposing for an OpenCL backend-specific function makes sense, and we could even return an OpenCL type (instead of a `vec<uint, 4>`) if we did that.

This does raise two interesting questions, though:

1. Should we say that the number of bits in an implementation's mask is defined in some macro somewhere?
  This would allow us to default it to 128-bit for portability, but maybe override it if the user wants smaller (or larger) groups.

2. Should we change the insert/extract functions to use something that isn't a `vec`?
  `vec` is deprecated anyway, and there's a weaker argument for it being a `vec` without the OpenCL/SPIR-V connection.

@rolandschulz, any thoughts here?

Reply:
> Should we change the insert/extract functions to use something that isn't a vec?

Maybe they could use `marray` like:

```
void insert_bits(const marray<bool, group_mask_bits> &bits, id<1> pos = 0)
```

where `group_mask_bits` is a `constexpr` value defined by the implementation.

Reply:
Does SYCL 2020 say anything about whether `marray<bool, N>` is optimized to store the `bool`s as individual bits?  If not, I think keeping this as `marray<uint32_t, group_mask_bits/sizeof(uint32_t)>`  might give better performance.

I also think we should keep the input type to `insert_bits` as a template parameter.  The intent is to allow people to write things like:
```c++
uint32_t lowest_32_bits = 0xFFFFFFFF;
mask.insert_bits<uint32_t>(lowest_32_bits, 0);
```
or
```c++
uint32_t lowest_32_bits = mask.extract_bits<uint32_t>(0);
```
...rather than force the user to ferry everything via an intermediate type.

Reply:
> Does SYCL 2020 say anything about whether marray<bool, N> is optimized to store the bools as individual bits? If not, I think keeping this as marray<uint32_t, group_mask_bits/sizeof(uint32_t)> might give better performance.

Hmm, good point.  Originally, I was thinking that implementations would likely store `marray<bool, N>` as a bit array, but now I think the spec might prevent an implementation from doing that.  See [Section 4.14.3.3](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#memory-layout-and-alignment.marray):

> The elements of an instance of the marray class template as if stored in `std::array<dataT, numElements>`.

I think that prevents an implementation from packing `bool` elements into a bit array.

Reply:
I like the new wording, so I'm resolving this conversation.